### PR TITLE
privilege: fix user with `%` hostname can not show grants.

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve?

#15524 didn't fix this case:

```
create user u1;
// login as u1
show grants;
```
Authhostname of u1 is `%`, hostname of `u1` is `127.0.0.1`.  `ShowGrants` checks hostname in executor which cause show grants fail.

### What is changed and how it works?
if user in `ShowGrants` is `nil`, copy session variable user to executor.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- None

### CODE REVIEW

File: executor/show_test.go [hunk 1]
```diff
@@ -150,6 +150,14 @@ func (s *testSuite5) TestShowGrantsPrivilege(c *C) {
 	tk1.Se = se
 	err = tk1.QueryToErr("show grants for root")
 	c.Assert(err.Error(), Equals, executor.ErrDBaccessDenied.GenWithStackByArgs("show_grants", "%", mysql.SystemDB).Error())
+	// Test show grants for user with auth host name `%`.
+	tk2 := testkit.NewTestKit(c, s.store)
+	se2, err := session.CreateSession4Test(s.store)
+	c.Assert(err, IsNil)
+	c.Assert(se2.Auth(&auth.UserIdentity{Username: "show_grants", Hostname: "127.0.0.1", AuthUsername: "show_grants", AuthHostname: "%"}, nil, nil), IsTrue)
+	tk2.Se = se2
+	err = tk2.QueryToErr("show grants")
```
Comment (jackysp) [inline executor/show_test.go]: why not use `MustQuery`?

**Comment by tiancaiamao** [comment]
LGTM

File: executor/show_test.go (hunk 1 above)
Comment (imtbkcat) [reply executor/show_test.go]: fixed

**Pull request merged**